llvm/lib/Linker/LinkModules.cpp

```diff
@@ -497,6 +501,23 @@ bool ModuleLinker::run() {
   for (Function &GV : llvm::make_early_inc_range(DstM))
     dropReplacedComdat(GV, ReplacedDstComdats);
 
+  if (!NonPrevailingComdats.empty()) {
+    DenseSet<GlobalObject *> AliasedGlobals;
+    for (auto &GA : SrcM->aliases())
+      if (GlobalObject *GO = GA.getAliaseeObject(); GO && GO->getComdat())
+        AliasedGlobals.insert(GO);
+    for (const Comdat *C : NonPrevailingComdats) {
+      SmallVector<GlobalObject *> ToUpdate;
+      for (GlobalObject *GO : C->getUsers())
+        if (GO->hasPrivateLinkage() && !AliasedGlobals.contains(GO))
```

Comment:
I think it's more correct to use `hasLocalLinkage` instead of `hasPrivateLinkage`. Looking at the IRMover call sites, that's what's used there.

Reply:
This change would break the `Linker/comdat-refer-to-discarded.ll` test. Originally, it was `Linker/comdat13.ll` in https://github.com/llvm/llvm-project/commit/baa3bf8f76568de6198613fb2ba7695a76c29109, which stated that the duplication of `internal` globals "matches the expectations of COFF linkers`. Since no high-level example was provided, it is now quite difficult, at least for me, to judge whether that behavior is correct or not, so I would prefer not to change it in my patch.

Reply:
Looking at the test and the history, it is still valid. I think the bug here is that these comdats referring to COFF internal linkage entities are considered `NonPrevailingComdats`. The proper fix for that is probably in some other part of the code.

Reply:
I think we should go ahead and land this as is. I'll file an issue about the logic bug I think I see.

Reply:
`for (auto &GA : SrcM->aliases())` is not tested. Add some tests?

Reply:
> `for (auto &GA : SrcM->aliases())` is not tested. Add some tests?

This is covered in `Linker/comdat-any.ll`, case 3.